### PR TITLE
hid: Add SixAxis sensor support

### PR DESCRIPTION
**Pull request by endrift**
Adds the structs that SixAxis sensor states use in the controller structs. ~~I don't have code for reading them yet since I'm not sure how to select which layout I should be using yet.~~

### CODE REVIEW

File: nx/include/switch/services/hid.h [hunk 1]
```diff
@@ -506,11 +513,41 @@ typedef struct HidControllerLayout
 } HidControllerLayout;
 static_assert(sizeof(HidControllerLayout) == 0x350, "Hid controller layout structure has incorrect size");
 
+typedef struct HidControllerSixAxisHeader
+{
+    u64 unk;
```
Comment (yellows8) [inline nx/include/switch/services/hid.h]: timestamp

**Comment by endrift** [comment]
Added sensor value reading.

File: nx/source/services/hid.c [hunk 2]
```diff
@@ -227,6 +230,32 @@ void hidScanInput(void) {
 
         g_controllerDown[i] = (~g_controllerOld[i]) & g_controllerHeld[i];
         g_controllerUp[i] = g_controllerOld[i] & (~g_controllerHeld[i]);
+
+        u32 type = g_controllerHeaders[i].type;
```
Comment (yellows8) [inline nx/source/services/hid.c]: Only run this when SixAxis is activated.

Comment (endrift) [reply nx/source/services/hid.c]: How would you suggest I go about detecting that? I can add a check to not do the memcpy when the SixAxis is deactivated but the rest of the code can't easily do that without bookkeeping that could get out of sync.

If the former then I should also add cleanup when stopping a SixAxis sensor to make sure you don't read out old states.

Comment (yellows8) [reply nx/source/services/hid.c]: Set a flag in a table for the specified controller in the enable/disable funcs.

File: nx/source/services/hid.c [hunk 3]
```diff
@@ -394,6 +425,33 @@ void hidJoystickRead(JoystickPosition *pos, HidControllerID id, HidControllerJoy
...
+    if (id == CONTROLLER_P1_AUTO) return hidSixAxisSensorValuesRead(values, g_controllerP1AutoID, num_entries);
+
+    if (id < 0 || id > 9) {
+        memset(values, 0, sizeof(SixAxisSensorValues) * num_entries);
+        return;
+    }
+
+    rwlockReadLock(&g_hidLock);
+    entry = g_sixaxisLayouts[id].header.latestEntry;
+    for (i = num_entries; i--;) {
```
[the remark this answers is not in the document]
Comment (endrift) [reply nx/source/services/hid.c]: Added, though I'm not sure if it's quite how you wanted it to abort.

File: nx/source/services/hid.c [hunk 4]
```diff
@@ -394,6 +425,46 @@ void hidJoystickRead(JoystickPosition *pos, HidControllerID id, HidControllerJoy
...
+    if (num_entries > g_sixaxisLayouts[id].header.maxEntryIndex + 1)
+        num_entries = g_sixaxisLayouts[id].header.maxEntryIndex + 1;
+
+    entry = g_sixaxisLayouts[id].header.latestEntry + 1 - num_entries;
+    if (entry < 0)
+        entry += g_sixaxisLayouts[id].header.maxEntryIndex + 1;
+
+    u64 timestamp = 0;
+    for (i = 0; i < num_entries; i++) {
+        if (timestamp && g_sixaxisLayouts[id].entries[entry].timestamp - timestamp != 1) {
```
Comment (yellows8) [inline nx/source/services/hid.c]: Remove {}

File: nx/source/services/hid.c [hunk 5]
```diff
@@ -394,6 +425,46 @@ void hidJoystickRead(JoystickPosition *pos, HidControllerID id, HidControllerJoy
...
+
+    if (id == CONTROLLER_P1_AUTO) return hidSixAxisSensorValuesRead(values, g_controllerP1AutoID, num_entries);
+
+    if (id < 0 || id > 9) {
+        return 0;
+    }
+
+
+    rwlockReadLock(&g_hidLock);
+    if (num_entries > g_sixaxisLayouts[id].header.maxEntryIndex + 1)
```
[the remark this answers is not in the document]
Comment (endrift) [reply nx/source/services/hid.c]: maxEntryIndex can be < 0x10 for the first few timestamps, but numEntries is always 0x11. I'm not sure the change you're suggesting is correct.

**Comment by yellows8** [comment]
Check g_sixaxisEnabled in hidSixAxisSensorValuesRead. Locking is missing in hidStartSixAxisSensor/hidStopSixAxisSensor.

File: nx/source/services/hid.c [hunk 6]
```diff
@@ -19,11 +19,13 @@ static HidMouseEntry g_mouseEntry;
 static HidKeyboardEntry g_keyboardEntry;
 static HidControllerHeader g_controllerHeaders[10];
 static HidControllerInputEntry g_controllerEntries[10];
+static HidControllerSixAxisLayout g_sixaxisLayouts[10];
 
 static u64 g_mouseOld, g_mouseHeld, g_mouseDown, g_mouseUp;
 static u64 g_keyboardModOld, g_keyboardModHeld, g_keyboardModDown, g_keyboardModUp;
 static u32 g_keyboardOld[8], g_keyboardHeld[8], g_keyboardDown[8], g_keyboardUp[8];
 static u64 g_controllerOld[10], g_controllerHeld[10], g_controllerDown[10], g_controllerUp[10];
+static bool g_sixaxisEnabled[10];
```
Comment (yellows8) [inline nx/source/services/hid.c]: memset for this is missing in hidReset.

File: nx/include/switch/services/hid.h [hunk 7]
```diff
@@ -564,6 +606,8 @@ void* hidGetSharedmemAddr(void);
 
 void hidSetControllerLayout(HidControllerID id, HidControllerLayoutType layoutType);
 HidControllerLayoutType hidGetControllerLayout(HidControllerID id);
+u32 hidGetControllerType(HidControllerID id);
```
Comment (yellows8) [inline nx/include/switch/services/hid.h]: Change output to HidControllerType.

**Comment by endrift** [comment]
Updated and squashed everything.

**Pull request merged**